core/src/main/java/org/apache/calcite/rel/metadata/RelMdUtil.java

```diff
@@ -731,19 +734,22 @@ public static Double getJoinRowCount(RelMetadataQuery mq, Join join,
         return max;
       }
     }
-    double product = left * right;
-
-    return product * mq.getSelectivity(join, condition);
-  }
 
-  /** Returns an estimate of the number of rows returned by a semi-join. */
-  public static Double getSemiJoinRowCount(RelMetadataQuery mq, RelNode left,
-      RelNode right, JoinRelType joinType, RexNode condition) {
-    final Double leftCount = mq.getRowCount(left);
-    if (leftCount == null) {
-      return null;
+    double product = left * right;
+    double selectivity = mq.getSelectivity(join, condition);
+    double innerRowCount = product * selectivity;
+    switch (join.getJoinType()) {
+    case INNER:
+      return innerRowCount;
+    case LEFT:
+      return left * (1D - selectivity) + innerRowCount;
+    case RIGHT:
+      return right * (1D - selectivity) + innerRowCount;
```

Comment:
(1D - selectivity) * left/right to estimate the number of `row,null` tuple makes sense (although it replies on a good estimation on `selectivity`, which usually is a hard engineering problem :-))

Reply:
I agree. IMHO the proposed formula is far from perfect and depends on how good (or bad) is the selectivity estimation, but I think it is a bit better than the current formula.

Reply:
agreed. I think your change indeed is better.